README.md
# halculator
A simple RPN calculator written in Haskell
---
![status](https://travis-ci.org/tim6her/halculator.svg?branch=master)

Ever wanted to find the expected number of heads when 
tossing 5 coins, i. e., computing

![expected value](http://latex.univie.ac.at/?%5Cmathbb%20E%5Cleft%5B%5Cmathrm%7BBinom%7D%5Cleft%285,%20%5Cfrac%7B1%7D%7B2%7D%5Cright%29%5Cright%5D%20%3D%20%5Csum_%7Bi%3D0%7D%5E5%20i%20%7B5%20i%20%5Cchoose%20i%7D%20%5Cleft%28%5Cfrac%7B1%7D%7B2%7D%5Cright%29%5E5),

using an [RPN](https://en.wikipedia.org/wiki/Reverse_Polish_notation) calculator? Now you can!

![screen2](docs/screen2.gif)

### But wait there's more

Currently the following functions are implemented:

* Operators
  * `+`
  * `-`
  * `*`
  * `/`
  * `^`
* Misc
  * `sum` (sum of all elements in stack)
  * `prod` (product of all elements in stack)
  * `!` (factorial)
  * `nCr` (n choose r)
  * `abs`
  * `..` (range between two values)
* Constants
  * `pi`
  * `e`
* Functions with Exponents
  * `exp`
  * `log` (logarithm with respect to Euler constant `e`)
  * `ln` (alias for `log`)
  * `sqrt`
  * `logBase` (Logarithm of `x` with respect to base `y`)
* Trigonomitry and Hyperbolics
  * `sin`
  * `cos`
  * `tan`
  * `asin`
  * `acos`
  * `atan`
  * `sinh`
  * `cosh`
  * `tanh`
  * `asinh`
  * `acosh`
  * `atanh`
* Control
  * `swp` (swap the first two elements in stack)
  * `cpy` (copy first element)
  * `rot` (rotate stack one element to the left)
  * `rotl` (alias for `rot`)
  * `rotr` (rotate stack one element to the right)
  * `clr` (clear stack)

For a documentation of the implementation please visit https://tim6her.github.io/halculator/index.html

ChangeLog.md
# Revision history for halculator

## 0.1.0.0  -- 2017-02-13

* First version. Released on an unsuspecting world.


test/doctests.hs
import Test.DocTest

main = doctest ["Main.hs"]

calc.hs
data Stack a = EmptyStack | Push a (Stack a)
    deriving (Read, Eq, Ord) 

instance (Show a) => Show (Stack a) where
    show = (++ " <-") . show . reverse . toList 

pop :: Stack b -> (Maybe b, Stack b)
pop  EmptyStack = (Nothing, EmptyStack)
pop (Push x stack) = (Just x, stack)

popN :: (Integral a, Eq b) => a -> Stack b -> 
                              (Maybe [b], Stack b)
popN _ EmptyStack = (Nothing, EmptyStack)
popN 1 (Push x stack) = (Just [x], stack)
popN n (Push x stack)
    | n < 1 = error "Can't return less than one item"
    | otherwise = if may == Nothing
                  then (Nothing, EmptyStack)
                  else let (Just xs) = may
                       in (Just (x:xs), stack')
    where (may, stack') = popN (n - 1) stack

push :: b -> Stack b -> Stack b
push x stack = Push x stack

showFst :: (Show b) => Stack b -> String
showFst EmptyStack = "Nothing"
showFst (Push x _) = show x

toList :: Stack b -> [b]
toList EmptyStack = []
toList (Push x stack) = x:toList stack


main = calc EmptyStack

calc :: Stack Float -> IO ()
calc stack = do
    putStr "In: "
    comm <- getLine
    if comm == "show"
        then do
            print stack
            calc stack 
        else do
            stack' <- getStack stack $ pars stack comm
            putStr "Out: "
            putStrLn $ showFst stack'
            calc stack'

getStack :: (Eq a) => Stack a -> Maybe (Stack a) -> 
                      IO (Stack a)
getStack stack Nothing = do 
    putStrLn "Can't perform operation"
    return stack
getStack _ (Just stack') = return stack'

pars :: Stack Float -> String -> Maybe (Stack Float)
pars (Push x (Push y stack)) "+" = Just $ push (y + x) stack
pars (Push x (Push y stack)) "-" = Just $ push (y - x) stack
pars (Push x (Push y stack)) "*" = Just $ push (y * x) stack
pars (Push x (Push y stack)) "/" = Just $ push (y / x) stack
pars (Push x (Push y stack)) "^" = Just $ push (y ** x) stack
pars (Push x stack) "!" = Just $ push (fak x) stack
pars (Push x stack) "abs" = Just $ push (abs x) stack
pars stack "sum" = Just $ push (sum $ toList stack) EmptyStack
pars (Push x (Push y stack)) "swp" = Just $ push y (push x stack)
pars _ "clr" = Just EmptyStack
pars stack comm
    | contNum comm = Just $ push (read comm :: Float) stack
    | otherwise = Nothing

contNum :: String -> Bool
contNum comm = any (\ n -> n `elem` comm) ['0'..'9']

fak :: Float -> Float
fak x
    | x' == 0.0 = 1.0
    | x' == 1.0 = 1.0
    | x < 0 = error "Negativ factorial"
    | otherwise = x' * (fak $ x' - 1.0)
    where x' = (fromIntegral (truncate x :: Int) :: Float)


halc.hs
{-|
Module      : Halc
Description : The backbone of the executable
Copyright   : (c) Tim B. Herbstrith, 2017
License     : MIT
Maintainer  : tim.herbstrith@icoud.com
Stability   : experimental
Portability : POSIX

A simple RPN calculator written in Haskell
-}
module Halc where

import System.IO

-- * Functions for user interactions

-- | All numerals for computations are stored in
-- a stack of arbitrary size
type Stack = [Float]

-- | Starts the main recursion with empty stack
main = calc []

-- | Gets command and updates stack according
-- to user input.
--
-- The command "show" will print the stack.  All other
-- inputs will be passed to parsing.  At the end 'calc'
-- is called with the updated stack.
calc :: Stack -> IO ()
calc stack = do
    putStr "In: "
    hFlush stdout
    comm <- getLine
    if comm == "show"
    then do
        putStrLn $ show (reverse stack) ++ " <-"
        calc stack 
    else do
        stack' <- updateStack stack $
                        parsing stack $
                        words comm
        putStr "Out: "
        putStrLn $ showHead stack'
        calc stack'

-- | Returns string representation of the first item
-- on the stack.  If the stack is empty, the string
-- "Nothing" is returned.
showHead :: Stack -> String
showHead [] = "Nothing"
showHead (x:stack) = show x

-- | Safely updates the stack by checking first if
-- 'Maybe' 'Stack' is 'Nothing' the user will be
-- informed and the old stack is returned.
-- Otherwise the stack is extracted and returned.
updateStack :: Stack        -- ^ The old stack
            -> Maybe Stack  -- ^ Hopefully the new one
            -> IO Stack
updateStack stack Nothing = do 
    putStrLn "Can't perform operation"
    return stack
updateStack _ (Just stack') = return stack'

-- * Functions for parsing

-- | Takes a list of commands and applys them
-- sequencially to the stack.
--
-- If a computation fails, 'Nothing' is returned.
--
-- === Examples:
-- >>> parsing [] $ words "3 2 ^ 4 2 ^ + sqrt"
-- Just [5.0]
-- >>> parsing [2.0] ["/"] -- Too few arguments on stack!
-- Nothing
parsing :: Stack -> [String] -> Maybe Stack
parsing stack = foldl pars' mstack
    where mstack = Just stack
          pars' = \mstack' com -> do
                      stack' <- mstack'
                      pars stack' com

-- | Parses command and applys the corresponding
-- function to the first elements of the stack.
--
-- If the command is not recognised or there are
-- too few numbers on the stack, 'Nothing' is
-- returned.
--
-- === Examples:
-- >>> let stack = [-6.0, 2.0]
-- >>> pars stack "*"
-- Just [-12.0]
-- >>> pars stack "abs"
-- Just [6.0,2.0]
-- >>> pars stack ".."
-- Just [-6.0,-5.0,-4.0,-3.0,-2.0,-1.0,0.0,1.0,2.0]
-- >>> pars stack "katze" -- I can't imagine such a command exists
-- Nothing
pars :: Stack -> String -> Maybe Stack
-- Operators
pars (x:y:stack) "+" = Just $ (y + x) : stack
pars (x:y:stack) "-" = Just $ (y - x) : stack
pars (x:y:stack) "*" = Just $ (y * x) : stack
pars (x:y:stack) "/" = Just $ (y / x) : stack
pars (x:y:stack) "^" = Just $ (y ** x) : stack
-- Misc
pars stack "sum" = Just $ (sum stack) : []
pars stack "prod" = Just $ (foldl (*) 1 stack) : []
pars (x:stack) "!" = Just $ (fac x) : stack
pars (x:y:stack) "nCr" = Just $ (nCr y x) : stack
pars (x:stack) "abs" = Just $ (abs x) : stack
pars (x:y:stack) ".." = let a = truncate y
                            b = truncate x
                            a' = a + signum (b - a)
                        in Just $ map fromIntegral 
                                          (reverse [a,a'..b])
                                  ++ stack
-- Constants
pars stack "pi" = Just $ pi : stack
pars stack "e" = Just $ (exp 1.0) : stack
-- Functions with Exponents
pars (x:stack) "exp" = Just $ (exp x) : stack
pars (x:stack) "log" = Just $ (log x) : stack
pars stack "ln" = pars stack "log"
pars (x:stack) "sqrt" = Just $ (sqrt x) : stack
pars (x:y:stack) "logBase" = Just $ (logBase y x) : stack
-- Trigonomitry and Hyperbolics
pars (x:stack) "sin" = Just $ (sin x) : stack
pars (x:stack) "cos" = Just $ (cos x) : stack
pars (x:stack) "tan" = Just $ (tan x) : stack
pars (x:stack) "asin" = Just $ (asin x) : stack
pars (x:stack) "acos" = Just $ (acos x) : stack
pars (x:stack) "atan" = Just $ (atan x) : stack
pars (x:stack) "sinh" = Just $ (sinh x) : stack
pars (x:stack) "cosh" = Just $ (cosh x) : stack
pars (x:stack) "tanh" = Just $ (tanh x) : stack
pars (x:stack) "asinh" = Just $ (asinh x) : stack
pars (x:stack) "acosh" = Just $ (acosh x) : stack
pars (x:stack) "atanh" = Just $ (atanh x) : stack
-- Control
pars (x:y:stack) "swp" = Just $ y : x : stack
pars stack@(x:xs) "cpy" = Just $ x : stack
pars stack@(x:xs) "rot" = Just $ (last stack) : (init stack)
pars stack "rotl" = pars stack "rot"
pars (x:stack) "rotr" = Just $ stack ++ [x]
pars _ "clr" = Just []
pars stack comm
    | contNum comm = Just $ (read comm) : stack
    | otherwise = Nothing

-- * Operators and Helper Functions

-- | Checks whether a string contains a numeral
contNum :: String -> Bool
contNum comm = any (\ n -> n `elem` comm) ['0'..'9']

-- | A @Float@ implementation for the factorial
--
-- === Note:
-- Floats will be truncated.
fac :: (RealFrac a, Eq a, Ord a) => a -> a
fac x
    | x' == 0.0 = 1.0
    | x' == 1.0 = 1.0
    | x < 0 = 0.0 / 0.0 -- NaN representation in GHC
    | otherwise = x' * (fac $ x' - 1.0)
    where x' = fromIntegral (truncate x :: Int)

-- | A 'Float' implementation for "n choose r"
--
-- === Note:
-- Floats will be truncated.
nCr :: Float -> Float -> Float
nCr n r = fromIntegral $ nCr' (truncate n) (truncate r)

-- | The backbone of 'nCr'
nCr' :: (Integral a) => a -> a -> a
nCr' n r
    | n < r || r < 0 = div 0 0
    | r > (div n 2) = nCr' n (n - r)
    | r == 0 = 1
    | r == 1 = n
    | otherwise = (nCr' (n - 1) (r - 1)) + (nCr' (n - 1) r)


Main.hs
{-|
Module      : Main
Description : The backbone of the executable
Copyright   : (c) Tim B. Herbstrith, 2017
License     : MIT
Maintainer  : Please contact me via GitHub!
Stability   : experimental
Portability : POSIX

A simple RPN calculator written in Haskell
-}
module Main where

import System.IO

-- * Functions for user interactions

-- | All numerals for computations are stored in
-- a stack of arbitrary size
type Stack = [Float]

-- | Starts the main recursion with empty stack
main = calc []

-- | Gets command and updates stack according
-- to user input.
--
-- The command "show" will print the stack.  All other
-- inputs will be passed to parsing.  At the end 'calc'
-- is called with the updated stack.
calc :: Stack -> IO ()
calc stack = do
    putStr "In: "
    hFlush stdout
    comm <- getLine
    if comm == "show"
    then do
        putStrLn $ show (reverse stack) ++ " <-"
        calc stack 
    else do
        stack' <- updateStack stack $
                        parsing stack $
                        words comm
        putStr "Out: "
        putStrLn $ showHead stack'
        calc stack'

-- | Returns string representation of the first item
-- on the stack.  If the stack is empty, the string
-- "Nothing" is returned.
showHead :: Stack -> String
showHead [] = "Nothing"
showHead (x:stack) = show x

-- | Safely updates the stack by checking first if
-- 'Maybe' 'Stack' is 'Nothing' the user will be
-- informed and the old stack is returned.
-- Otherwise the stack is extracted and returned.
updateStack :: Stack        -- ^ The old stack
            -> Maybe Stack  -- ^ Hopefully the new one
            -> IO Stack
updateStack stack Nothing = do 
    putStrLn "Can't perform operation"
    return stack
updateStack _ (Just stack') = return stack'

-- * Functions for parsing

-- | Takes a list of commands and applys them
-- sequencially to the stack.
--
-- If a computation fails, 'Nothing' is returned.
--
-- === Examples:
-- >>> parsing [] $ words "3 2 ^ 4 2 ^ + sqrt"
-- Just [5.0]
-- >>> parsing [2.0] ["/"] -- Too few arguments on stack!
-- Nothing
parsing :: Stack -> [String] -> Maybe Stack
parsing stack = foldl pars' mstack
    where mstack = Just stack
          pars' = \mstack' com -> do
                      stack' <- mstack'
                      pars stack' com

-- | Parses command and applys the corresponding
-- function to the first elements of the stack.
--
-- If the command is not recognised or there are
-- too few numbers on the stack, 'Nothing' is
-- returned.
--
-- === Examples:
-- >>> let stack = [-6.0, 2.0] :: Stack
-- >>> pars stack "*"
-- Just [-12.0]
-- >>> pars stack "abs"
-- Just [6.0,2.0]
-- >>> pars stack ".."
-- Just [-6.0,-5.0,-4.0,-3.0,-2.0,-1.0,0.0,1.0,2.0]
-- >>> pars stack "katze" -- I can't imagine such a command exists
-- Nothing
pars :: Stack -> String -> Maybe Stack
-- Operators
pars (x:y:stack) "+" = Just $ (y + x) : stack
pars (x:y:stack) "-" = Just $ (y - x) : stack
pars (x:y:stack) "*" = Just $ (y * x) : stack
pars (x:y:stack) "/" = Just $ (y / x) : stack
pars (x:y:stack) "^" = Just $ (y ** x) : stack
-- Misc
pars stack "sum" = Just $ (sum stack) : []
pars stack "prod" = Just $ (foldl (*) 1 stack) : []
pars (x:stack) "!" = Just $ (fac x) : stack
pars (x:y:stack) "nCr" = Just $ (nCr y x) : stack
pars (x:stack) "abs" = Just $ (abs x) : stack
pars (x:y:stack) ".." = let a = truncate y
                            b = truncate x
                            a' = a + signum (b - a)
                        in Just $ map fromIntegral 
                                          (reverse [a,a'..b])
                                  ++ stack
-- Constants
pars stack "pi" = Just $ pi : stack
pars stack "e" = Just $ (exp 1.0) : stack
-- Functions with Exponents
pars (x:stack) "exp" = Just $ (exp x) : stack
pars (x:stack) "log" = Just $ (log x) : stack
pars stack "ln" = pars stack "log"
pars (x:stack) "sqrt" = Just $ (sqrt x) : stack
pars (x:y:stack) "logBase" = Just $ (logBase y x) : stack
-- Trigonomitry and Hyperbolics
pars (x:stack) "sin" = Just $ (sin x) : stack
pars (x:stack) "cos" = Just $ (cos x) : stack
pars (x:stack) "tan" = Just $ (tan x) : stack
pars (x:stack) "asin" = Just $ (asin x) : stack
pars (x:stack) "acos" = Just $ (acos x) : stack
pars (x:stack) "atan" = Just $ (atan x) : stack
pars (x:stack) "sinh" = Just $ (sinh x) : stack
pars (x:stack) "cosh" = Just $ (cosh x) : stack
pars (x:stack) "tanh" = Just $ (tanh x) : stack
pars (x:stack) "asinh" = Just $ (asinh x) : stack
pars (x:stack) "acosh" = Just $ (acosh x) : stack
pars (x:stack) "atanh" = Just $ (atanh x) : stack
-- Control
pars (x:y:stack) "swp" = Just $ y : x : stack
pars stack@(x:xs) "cpy" = Just $ x : stack
pars stack@(x:xs) "rot" = Just $ (last stack) : (init stack)
pars stack "rotl" = pars stack "rot"
pars (x:stack) "rotr" = Just $ stack ++ [x]
pars _ "clr" = Just []
pars stack comm
    | contNum comm = Just $ (read comm) : stack
    | otherwise = Nothing

-- * Operators and Helper Functions

-- | Checks whether a string contains a numeral
contNum :: String -> Bool
contNum comm = any (\ n -> n `elem` comm) ['0'..'9']

-- | A @Float@ implementation for the factorial
--
-- === Note:
-- Floats will be truncated.
--
-- === Example:
-- >>> fac 4.2
-- 24.0
fac :: (RealFrac a, Eq a, Ord a) => a -> a
fac x
    | x' == 0.0 = 1.0
    | x' == 1.0 = 1.0
    | x < 0 = 0.0 / 0.0 -- NaN representation in GHC
    | otherwise = x' * (fac $ x' - 1.0)
    where x' = fromIntegral (truncate x :: Int)

-- | A 'Float' implementation for "n choose r"
--
-- === Note:
-- Floats will be truncated.
nCr :: Float -> Float -> Float
nCr n r = fromIntegral $ nCr' (truncate n) (truncate r)

-- | The backbone of 'nCr'
nCr' :: (Integral a) => a -> a -> a
nCr' n r
    | n < r || r < 0 = div 0 0
    | r > (div n 2) = nCr' n (n - r)
    | r == 0 = 1
    | r == 1 = n
    | otherwise = (nCr' (n - 1) (r - 1)) + (nCr' (n - 1) r)
